File: src/types/workflow.ts
/**
 * Shared workflow contract types (B5 frontend hygiene).
 *
 * Single source of truth for the pilot path
 * `upload -> /tasks/{id} -> /processes -> /analysis/{id} -> /docx`,
 * mirroring `docs/coordination/api-contract.md` and the backend
 * `app/schemas/workflow.py` shapes. Pages must import from here
 * instead of redeclaring inline interfaces.
 */

/** `POST /api/v1/documents/upload` response. */
export interface UploadResponse {
  id: string;
  task_id?: string;
  taskStatusUrl?: string;
  status: string;
}

/**
 * `GET /api/v1/tasks/{task_id}` response.
 *
 * BL-014 identity rule: `task_id == document_id` by design — one
 * document owns exactly one pipeline (single `Analysis` per document).
 */
export interface TaskStatusResponse {
  task_id?: string;
  document_id?: string;
  status?: string;
  progress?: number;
  analysis_id?: string | null;
  status_detail?: string | null;
  error_message?: string | null;
}

/** `GET /api/v1/processes` item (dashboard source of truth). */
export interface ProcessItem {
  id: string;
  analysis_id?: string | null;
  title: string;
  status: string;
  created_at?: string;
  status_detail?: string | null;
}

/** `GET /api/v1/analysis/{id}` response (camel + snake tolerant). */
export interface AnalysisDetailResponse {
  id: string;
  document_id: string;
  documentName?: string;
  document_name?: string;
  title: string;
  summary: string;
  keyArguments?: string[];
  key_arguments?: string[];
  requests?: string[];
  laws?: string[];
  evidence?: unknown;
  defense_theses?: string[];
  generatedDefenseStrategy?: string;
  generated_defense_strategy?: string;
  docxDownloadUrl?: string | null;
  docx_download_url?: string | null;
}

/** Normalized analysis shape used by the analysis page UI. */
export interface NormalizedAnalysis {
  id: string;
  documentId: string;
  documentName: string;
  title: string;
  summary: string;
  keyArguments: string[];
  requests: string[];
  laws: string[];
  evidence: string[];
  defenseTheses: string[];
  generatedDefenseStrategy: string;
  docxDownloadUrl: string | null;
}

/** Structured `404 analysis-not-ready` detail from the backend. */
export interface AnalysisNotReadyDetail {
  message?: string;
  task_id?: string;
  status?: string;
  status_detail?: string;
}

/**
 * `GET /api/v1/templates` item (C1 template MVP).
 *
 * `unsupportedPlaceholders` names `{{roots}}` with no normalized analysis
 * context key — generation with such a template fails with 422.
 */
export interface UserTemplate {
  id: string;
  name: string;
  placeholders?: string[];
  unsupportedPlaceholders?: string[];
  created_at?: string;
}

/** Structured `422 incompatible-template` detail from the backend. */
export interface TemplateIncompatibilityDetail {
  message?: string;
  unsupported_placeholders?: string[];
  supported_keys?: string[];
}

/** `GET /api/v1/analysis/{id}/versions` item (C2 generation history). */
export interface GeneratedVersion {
  version: number;
  template_id?: string | null;
  created_at?: string;
  downloadUrl?: string | null;
}

const ACTIVE_STATUSES = new Set(['PENDING', 'PROCESSING', 'STARTED', 'RETRY']);
const SUCCESS_STATUSES = new Set(['SUCCESS', 'COMPLETED', 'DONE']);
const FAILURE_STATUSES = new Set(['FAILURE', 'FAILED', 'ERROR']);

export function normalizeWorkflowStatus(status?: string | null): string {
  return (status || 'PENDING').toUpperCase();
}

export function isActiveStatus(status?: string | null): boolean {
  return ACTIVE_STATUSES.has(normalizeWorkflowStatus(status));
}

export function isSuccessStatus(status?: string | null): boolean {
  return SUCCESS_STATUSES.has(normalizeWorkflowStatus(status));
}

export function isFailureStatus(status?: string | null): boolean {
  return FAILURE_STATUSES.has(normalizeWorkflowStatus(status));
}

File: backend/app/schemas/workflow.py
from datetime import datetime
from typing import Any

from pydantic import UUID4, BaseModel, ConfigDict, Field


class WorkflowBase(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class ProcessResponse(WorkflowBase):
    id: UUID4
    analysis_id: UUID4 | None = None
    title: str
    status: str
    created_at: datetime | None = None
    updated_at: datetime | None = None
    completed_at: datetime | None = None
    status_detail: str | None = None
    analysis_url: str | None = None
    docx_download_url: str | None = Field(default=None, alias="docxDownloadUrl")


class UploadSubmissionResponse(WorkflowBase):
    id: UUID4
    task_id: UUID4
    user_id: UUID4
    filename: str
    content_type: str
    status: str
    status_detail: str | None = None
    uploaded_at: datetime | None = None
    task_status_url: str = Field(alias="taskStatusUrl")
    analysis_id: UUID4 | None = None
    analysis_url: str | None = None
    docx_download_url: str | None = Field(default=None, alias="docxDownloadUrl")


class TaskStatusResponse(WorkflowBase):
    task_id: UUID4
    document_id: UUID4
    title: str
    status: str
    progress: int
    analysis_id: UUID4 | None = None
    status_detail: str | None = None
    error_message: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    completed_at: datetime | None = None
    analysis_url: str | None = None
    docx_download_url: str | None = Field(default=None, alias="docxDownloadUrl")


class AnalysisDetailResponse(WorkflowBase):

    id: UUID4
    document_id: UUID4
    document_name: str = Field(alias="documentName")
    title: str
    summary: str
    key_arguments: list[str] = Field(default_factory=list, alias="keyArguments")
    requests: list[str] = Field(default_factory=list)
    laws: list[str] = Field(default_factory=list)
    evidence: Any = None
    defense_theses: list[str] = Field(default_factory=list)
    status: str | None = None
    status_detail: str | None = None
    created_at: datetime | None = None
    completed_at: datetime | None = None
    generated_defense_strategy: str = Field(alias="generatedDefenseStrategy")
    docx_download_url: str | None = Field(default=None, alias="docxDownloadUrl")


class GeneratedVersionResponse(WorkflowBase):
    """One persisted DOCX generation (C2/BL-017 history)."""

    version: int
    template_id: UUID4 | None = None
    created_at: datetime | None = None
    download_url: str | None = Field(default=None, alias="downloadUrl")


class BatchUploadError(WorkflowBase):
    filename: str
    detail: str


class BatchUploadResponse(WorkflowBase):
    """Result of a multi-PDF batch upload (C3/BL-021)."""

    items: list[UploadSubmissionResponse] = Field(default_factory=list)
    errors: list[BatchUploadError] = Field(default_factory=list)
